### PR TITLE
BUGFIX: SD-632 use absolute position for tooltip

### PR DESCRIPTION
**Pull request by lanhq**
---

[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist

- [x] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [x] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [x] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [x] Migration guide (for major changes) is mentioned in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [x] Successful `extended test - examples`
- [x] Successful `extended test - storybook`
- [x] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)
- [ ] squashed commit

# Related PRs

- gdc-analytical-designer: https://github.com/gooddata/gdc-analytical-designer/pull/2480
- gdc-dashboards: https://github.com/gooddata/gdc-dashboards/pull/2288

# Related Jira tasks

- SD-632: https://jira.intgdc.com/browse/SD-632

### CODE REVIEW

**Comment by lanhq** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2542/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/86/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1939/ : SUCCESS in 15m 05s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4794/ : SUCCESS in 25m 41s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/85/ : SUCCESS in 11s

**Comment by lanhq** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1174.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/87/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1940/ : SUCCESS in 24m 07s

File: src/components/visualizations/styles/tooltip.scss [hunk 1]
```diff
@@ -180,3 +141,11 @@
         }
     }
 }
+
+// hide default highcharts svg tooltip when using html-based one
+.highcharts-tooltip-container {
+    g.highcharts-tooltip {
+        visibility: hidden;
+    }
+    z-index: 3005;
```
Comment (imnutz) [inline src/components/visualizations/styles/tooltip.scss]: Why 3005? This should be moved to _variables and add a comment to explain the number.

File: src/components/visualizations/styles/charts.scss [hunk 2]
```diff
@@ -37,6 +37,10 @@ $paging-button-small-size: 22px;
 }
 
 .highcharts-container {
+    // This make chart is visible inside the widget which have scroll bar
```
Comment (imnutz) [inline src/components/visualizations/styles/charts.scss]: This sentence is not correct grammatically

File: src/components/visualizations/styles/tooltip.scss (hunk 1 above)
Comment (lanhq) [reply src/components/visualizations/styles/tooltip.scss]: i just keep the old one.
https://github.com/gooddata/gooddata-react-components/pull/1174/files#diff-84b5bf43cbbf7caee4c0bdabb659f0efL23

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4802/ : SUCCESS in 7m 38s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/96/ : SUCCESS in 11s

File: src/components/visualizations/chart/highcharts/customConfiguration.ts [hunk 3]
```diff
@@ -260,30 +267,49 @@ function positionTooltip(chartType: any, stacking: any, labelWidth: any, labelHe
...
+    const leftOffset = pageXOffset + containerLeft - HIGHCHARTS_TOOLTIP_TOP_LEFT_OFFSET;
+    const topOffset = pageYOffset + containerTop - HIGHCHARTS_TOOLTIP_TOP_LEFT_OFFSET;
+
+    return {
+        x: showFullScreenTooltip() ? leftOffset : leftOffset + x,
+        y: topOffset + y,
+    };
+}
+
+const showFullScreenTooltip = () => {
```
[the remark this answers is not in the document]
Comment (lanhq) [reply src/components/visualizations/chart/highcharts/customConfiguration.ts]: - I changed from `showFullScreenTooltip` to `isShowFullScreenTooltip`.
- Does full screen mean large viewport size?
=> no. it mean, in view port <= 480. tooltip have size equal to chart container.

File: src/components/visualizations/chart/highcharts/customConfiguration.ts [hunk 4]
```diff
@@ -260,30 +267,49 @@ function positionTooltip(chartType: any, stacking: any, labelWidth: any, labelHe
...
-const showFullscreenTooltip = () => {
+export function calculateAbsolutePositionTooltip(
+    chartType: any,
+    stacking: any,
+    labelWidth: any,
+    labelHeight: any,
+    point: any,
+) {
+    const calculateRelatetivePosition = calculateRelativePositionTooltip.bind(this);
+    const { x, y } = calculateRelatetivePosition(chartType, stacking, labelWidth, labelHeight, point);
```
[the remark this answers is not in the document]
Comment (lanhq) [reply src/components/visualizations/chart/highcharts/customConfiguration.ts]: fixed

File: src/components/visualizations/chart/highcharts/customConfiguration.ts [hunk 5]
```diff
@@ -239,7 +240,13 @@ function getTooltipVerticalOffset(chartType: any, stacking: any, point: any) {
     return TOOLTIP_VERTICAL_OFFSET;
 }
 
-function positionTooltip(chartType: any, stacking: any, labelWidth: any, labelHeight: any, point: any) {
+export function calculateRelativePositionTooltip(
```
[the remark this answers is not in the document]
Comment (naibinh) [reply src/components/visualizations/chart/highcharts/customConfiguration.ts]: Now we have `calculateRelativePositionTooltip` and `calculateAbsolutePositionTooltip`
To me, it looks like we're going to do
```
if (something === true) {
    calculateRelativePositionTooltip();
} else {
    calculateAbsolutePositionTooltip();
}
```
But we're not.

So I suggest `getTooltipPositionInBody` and `getTooltipPositionInHighcharts`. They are more meaningful to me, not only we don't care about relative/absolute, but also explain us the steps to achieve tooltip position.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4803/ : SUCCESS in 7m 54s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/97/ : SUCCESS in 13s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4804/ : SUCCESS in 26m 41s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/98/ : SUCCESS in 11s

File: src/components/visualizations/chart/highcharts/customConfiguration.ts [hunk 6]
```diff
@@ -260,30 +270,55 @@ function positionTooltip(chartType: any, stacking: any, labelWidth: any, labelHe
...
+    const leftOffset = pageXOffset + containerLeft - HIGHCHARTS_TOOLTIP_TOP_LEFT_OFFSET;
+    const topOffset = pageYOffset + containerTop - HIGHCHARTS_TOOLTIP_TOP_LEFT_OFFSET;
+
+    return {
+        x: isTooltipShowedInFullScreen() ? leftOffset : leftOffset + x,
+        y: topOffset + y,
+    };
+}
+
+const isTooltipShowedInFullScreen = () => {
```
Comment (naibinh) [inline src/components/visualizations/chart/highcharts/customConfiguration.ts]: One last minor typo which is my bad.
`isTooltipShowedInFullScreen` -> `isTooltipShownInFullScreen`

Comment (lanhq) [reply src/components/visualizations/chart/highcharts/customConfiguration.ts]: updated

**Comment by lanhq** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4808/ : SUCCESS in 9m 06s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/102/ : SUCCESS in 9s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2545/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/103/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1943/ : SUCCESS in 13m 11s

**Comment by naibinh** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4811/ : SUCCESS in 9m 02s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/109/ : SUCCESS in 5s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4812/ : SUCCESS in 8m 55s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/110/ : SUCCESS in 7s

**Comment by ptquang86** [comment]
LGTM

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2547/ : SUCCESS in 14m 04s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4814/ : SUCCESS in 8m 58s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1210/ : SUCCESS in 21s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/1576/ : SUCCESS in 21m 59s

**Pull request merged**